include/libp2p/protocol/kademlia/impl/peer_routing_table_impl.hpp

```diff
@@ -50,39 +51,89 @@ namespace libp2p::protocol::kademlia {
       return std::memcmp(d1.data(), d2.data(), size) < 0;
     }
 
-    Hash256 hfrom;
+    Hash256 hfrom{};
   };
+
   /**
    * Single bucket which holds peers.
    */
-  class Bucket : public std::deque<BucketPeerInfo> {
+  class Bucket {
    public:
+    size_t size() const {
+      return peers_.size();
+    }
+
+    void append(const Bucket &bucket) {
+      peers_.insert(peers_.end(), bucket.peers_.begin(), bucket.peers_.end());
+    }
+
+    // sort bucket in ascending order by XOR distance from node_id
+    void sort(const NodeId &node_id) {
+      XorDistanceComparator cmp(node_id);
+      peers_.sort(cmp);
+    }
+
+    auto find(const peer::PeerId &p) {
+      return std::find_if(peers_.begin(), peers_.end(),
+                          [&p](const auto &i) { return i.peer_id == p; });
+    }
+
+    bool moveToFront(const PeerId &pid) {
+      auto it = find(pid);
+      if (it != peers_.end()) {
+        if (it != peers_.begin()) {
+          peers_.splice(peers_.begin(), peers_, it);
+        }
+        return false;
+      }
+      return true;
+    }
+
+    void emplaceToFront(const PeerId &pid, bool is_replaceable) {
+      peers_.emplace(peers_.begin(), pid, is_replaceable);
+    }
+
+    boost::optional<PeerId> removeReplaceableItem() {
+      boost::optional<PeerId> result;
+
+      for (auto it = peers_.rbegin(); it != peers_.rend(); ++it) {
+        if (it->is_replaceable) {
+          result = std::move(it->peer_id);
+          peers_.erase((++it).base());
+          break;
+        }
+      }
+
+      return result;
+    }
+
     void truncate(size_t limit) {
-      if (size() > limit) {
-        erase(std::next(begin(), limit), end());
+      if (limit == 0) {
+        peers_.clear();
+      } else if (peers_.size() > limit) {
+        size_t n = peers_.size() - limit;
+        for (size_t i = 0; i < n; ++i) {
+          peers_.pop_back();
```

Comment:
(not required)
could be just ```peers_.resize(limit);```

Reply:
there is std::list, maybe this is cheaper